src/datatable/datatable.jl

```diff
@@ -74,119 +74,111 @@ type DataTable <: AbstractDataTable
+        elseif minlen != maxlen
+            # recycle scalars
+            for i in 1:length(columns)
+                typeof(columns[i]) <: AbstractArray && continue
+                columns[i] = fill(columns[i], maxlen)
+                lengths[i] = maxlen
             end
-            if !equallengths
-                msg = "All columns in a DataTable must be the same length"
-                throw(ArgumentError(msg))
+            uls = unique(lengths)
+            if length(uls) != 1
+                strnames = string.(names(colindex))
+                estring = ["column length ($(uls[i])) for column(s) ($(join(strnames[find(uls .== u)], ", ")))"
+                           for (i,u) in enumerate(uls)]
+                throw(DimensionMismatch(join(estring, " is incompatible with ")))
             end
         end
-        if length(colindex) != ncols
-            msg = "Columns and column index must be the same length"
-            throw(ArgumentError(msg))
+        for (i,c) in enumerate(columns)
+            if isa(c, Range)
+                columns[i] = collect(c)
+            elseif !isa(c, AbstractVector)
+                if isa(c, AbstractArray)
+                    throw(DimensionMismatch("columns must be 1-dimensional"))
+                else
+                    columns[i] = [c]
+                end
+            else
+                columns[i] = c
+            end
         end
-        new(columns, colindex)
+        return new(columns, colindex)
     end
 end
 
 function DataTable(; kwargs...)
-    result = DataTable(Any[], Index())
-    for (k, v) in kwargs
-        result[k] = v
+    if length(kwargs) == 0
+        return DataTable(Any[], Index())
     end
-    return result
+    colnames = [Symbol(k) for (k,v) in kwargs]
+    columns = Any[v for (k,v) in kwargs]
+    DataTable(columns, Index(colnames))
 end
 
 function DataTable(columns::AbstractVector,
                    cnames::AbstractVector{Symbol} = gennames(length(columns)))
     return DataTable(convert(Vector{Any}, columns), Index(convert(Vector{Symbol}, cnames)))
 end
 
-
 # Initialize empty DataTable objects of arbitrary size
 function DataTable(t::Type, nrows::Integer, ncols::Integer)
     columns = Vector{Any}(ncols)
     for i in 1:ncols
-        columns[i] = NullableArray(t, nrows)
+        columns[i] = Vector{t}(nrows)
     end
     cnames = gennames(ncols)
     return DataTable(columns, Index(cnames))
 end
 
 # Initialize an empty DataTable with specific eltypes and names
-function DataTable(column_eltypes::Vector, cnames::Vector, nrows::Integer)
+function DataTable{T<:Type}(column_eltypes::AbstractVector{T}, cnames::AbstractVector{Symbol}, nrows::Integer)
     p = length(column_eltypes)
     columns = Vector{Any}(p)
     for j in 1:p
-        columns[j] = NullableArray(column_eltypes[j], nrows)
+        elty = column_eltypes[j]
+        columns[j] = elty <: Nullable ? NullableArray{eltype(elty)}(nrows) : Vector{elty}(nrows)
```

Comment:
Do you mean call `similar_nullable` when you say `(Categorical)NullableVector`? If so, I definitely agree, that would be the correct thing to do. But I'm not sure what you mean by `(Categorical)NullableVector` otherwise